crates/l2/sequencer/block_producer/payload_builder.rs

```diff
@@ -0,0 +1,307 @@
+) -> Result<PayloadBuildResult, BlockProducerError> {
+    let since = Instant::now();
+    let gas_limit = payload.header.gas_limit;
+
+    debug!("Building payload");
+    let mut context = PayloadBuildContext::new(payload, blockchain.evm_engine, store)?;
+
+    blockchain.apply_withdrawals(&mut context)?;
+    fill_transactions(blockchain.clone(), &mut context, store).await?;
+    blockchain.extract_requests(&mut context)?;
+    blockchain.finalize_payload(&mut context).await?;
+
+    let interval = Instant::now().duration_since(since).as_millis();
+    tracing::info!("[METRIC] BUILDING PAYLOAD TOOK: {interval} ms");
+    #[allow(clippy::as_conversions)]
+    if let Some(gas_used) = gas_limit.checked_sub(context.remaining_gas) {
+        let as_gigas = (gas_used as f64).div(10_f64.powf(9_f64));
+
+        if interval != 0 {
+            let throughput = (as_gigas) / (interval as f64) * 1000_f64;
+            tracing::info!(
+                "[METRIC] BLOCK BUILDING THROUGHPUT: {throughput} Gigagas/s TIME SPENT: {interval} msecs"
+            );
+        }
+    }
+
+    Ok(context.into())
+}
+
+/// Same as `blockchain::fill_transactions` but enforces that the `StateDiff` size  
+/// stays within the blob size limit after processing each transaction.
+pub async fn fill_transactions(
+    blockchain: Arc<Blockchain>,
+    context: &mut PayloadBuildContext,
+    store: &Store,
+) -> Result<(), BlockProducerError> {
+    // Two bytes for the len
+    let (mut acc_withdrawals_size, mut acc_deposits_size): (usize, usize) = (2, 2);
+    let mut acc_state_diff_size = 0;
+    let mut accounts_info_cache = HashMap::new();
+
+    let chain_config = store.get_chain_config()?;
+    let max_blob_number_per_block: usize = chain_config
+        .get_fork_blob_schedule(context.payload.header.timestamp)
+        .map(|schedule| schedule.max)
+        .unwrap_or_default()
+        .try_into()
+        .unwrap_or_default();
+
+    debug!("Fetching transactions from mempool");
+    // Fetch mempool transactions
+    let (mut plain_txs, mut blob_txs) = blockchain.fetch_mempool_transactions(context)?;
+    // Execute and add transactions to payload (if suitable)
+    loop {
+        // Check if we have enough gas to run more transactions
+        if context.remaining_gas < TX_GAS_COST {
+            debug!("No more gas to run transactions");
+            break;
+        };
+
+        // Check if we have enough space for the StateDiff to run more transactions
+        if acc_state_diff_size + TX_STATE_DIFF_SIZE > SAFE_BYTES_PER_BLOB {
+            debug!("No more StateDiff space to run transactions");
+            break;
+        };
+        if !blob_txs.is_empty() && context.blobs_bundle.blobs.len() >= max_blob_number_per_block {
+            debug!("No more blob gas to run blob transactions");
+            blob_txs.clear();
+        }
+        // Fetch the next transactions
+        let (head_tx, is_blob) = match (plain_txs.peek(), blob_txs.peek()) {
```

Comment:
I think we won't support blob txs in the L2, so maybe this is unnecessary

Reply:
Yeah, we almost certainly won't support them, but I would tackle this in a separate PR where we remove all blob txs support on the L2